5-AccessControl/2-call-api-groups/README.md

```diff
@@ -331,304 +393,218 @@ The client application React SPA has a [RouteGuard](./SPA/src/components/RouteGu
         }
-    }, [inProgress]);
 
-    return (
-        <>
-            { groupsData ? <GraphQuery groupsData={groupsData} /> : null}
-        </>
-    );
-};
+        if (result) {
+            fetchGroups(result.accessToken);
+        }
 
-export const Overage = () => {
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [instance, result, error, login, groupsData]);
 
-    const authRequest = {
-        ...loginRequest
-    };
+    if (error) {
+        return <div>Error: {error.message}</div>;
+    }
 
-    return (
-        <MsalAuthenticationTemplate
-            interactionType={InteractionType.Redirect}
-            authenticationRequest={authRequest}
-        >
-            <OverageContent />
-        </MsalAuthenticationTemplate>
-    )
+    return <>{groupsData ? <GraphQuery groupsData={groupsData} /> : null} </>;
 };
 ```
 
-#### Node.js handleOverage middleware
+#### Express custom handleOverage middleware
 
-Similar to the React **RouteGuard** component above, in Node.js **routeGuard** middleware we are checking whether the token for the user has the `_claim_names` claim, which indicates that the user has too many group memberships and thus overage occurs.
+Similar to the React **RouteGuard** component above, in custom Express.js middleware [routeGuard.js](./API/auth/guard.js) we are checking whether the token for the user has the `_claim_names` claim, which indicates that the user has too many group memberships and thus overage has occurred.
 
 ```javascript
-const routeGuard = (accessMatrix) => {
+const routeGuard = (accessMatrix, cache) => {
     return async (req, res, next) => {
         if (req.authInfo.groups === undefined) {
-            if (req.authInfo['_claim_names'] || req.authInfo['_claim_sources']) {
-                return handleOverage(req, res, next);
-            } else {
-                return res.status(403).json({ error: 'No group claim found!' });
+            if (hasOverageOccurred(req.authInfo)) {
+                return handleOverage(req, res, next, cache);
             }
+
+            return res.status(403).json({ error: 'No group claim found!' });
         } else {
-            // check for group IDs
+            if (!hasRequiredGroups(accessMatrix, req.path, req.method, req.authInfo['groups'])) {
+                return res.status(403).json({ error: 'User does not have the group, method or path' });
+            }
         }
 
         next();
-    }
-}
-
-module.exports = routeGuard;
+    };
+};
 ```
 
-If overage occurs, we initiate a call to MS Graph API's `https://graph.microsoft.com/v1.0/me/memberOf` endpoint to query the full list of groups that the user belongs to. Finally we check for the designated `groupID` among this list.
+If overage occurs, we initiate a call to MS Graph API's `https://graph.microsoft.com/v1.0/me/memberOf` endpoint to query the full list of groups that the user belongs to. Finally we check for the designated `groupID`s among this list.
```

Comment:
>memberOf [](http://example.com/codeflow?start=93&length=8)

hey, we have been advised to call getMemberGroups as its more performant

Reply:
done